### PR TITLE
Fixed incidence angle for SAC header

### PR DESCRIPTION
**Pull request by damhuonglan**
### What does this PR do?
Fixed incidence angle for SAC header such that the result is angles from the vertical axis instead of horizontal; Added test case to check that the CMPINC field in the header of the written SAC file;

### Relevant Issues?
#469

### Checklist
- [ ] This PR is not directly related to an existing issue (which has no PR yet).
- [ ] All tests pass.
- [ ] Any new or changed features have are documented.
- [ ] Changes have been added to `CHANGELOG.txt` .
- [ ] First time contributors have added your name to `CONTRIBUTORS.txt` .

### CODE REVIEW

**Comment by hrotman-pic** [comment]
Thank you for creating this PR.  I tested it, and it meets the requested behavior of CMPINC=90 for the horizontal channels, but unfortunately does not meet the requested behavior for CMPINC of the vertical channel.  For the vertical channel, I think ph5toms may need to look at the orientation/description_s field in the receiver table, and if orientation/description_s=Z, I believe SAC dip = dip + 90 will give the requested CMPINC behavior for the vertical channel.  For most sensors used in PH5 experiments, orientation/dip/value_f=90 for the vertical channel, so this would result in 90 + 90 = 180 CMPINC on the SAC header.  For sensors with the opposite convention, orientation/dip/value_f=-90 for the vertical channel, so -90 + 90 = 0 CMPINC on the SAC header.

**Review by hrotman-pic — APPROVED** [review]
I have tested this with the experiment on which it was originally reported (Fairfield SEG-D), as well as data from RT130, Texan, and SmartSolo experiments.  This is giving the expected CMPINC value for the vertical channel on PH5 experiments now, as well.  Thank you for making the additional changes.  The examples of CMPINC from testing, and reference CMPINC from SAC files using FDSNWS from two SEED experiments, are all below.

### From 20-001 (Fairfield SEG-D)
SAC> lh CMPINC


  ###### FILE: XE.221..DP1.2020-10-27T01.00.00.sac - 1
 -----------------------------------------

     CMPINC = 9.000000e+01

  ###### FILE: XE.221..DP2.2020-10-27T01.00.00.sac - 2
 -----------------------------------------

     CMPINC = 9.000000e+01

  ###### FILE: XE.221..DPZ.2020-10-27T01.00.00.sac - 3
 -----------------------------------------

     CMPINC = 1.800000e+02



### From 19-015 (Fairfield SEG-D)
SAC> lh CMPINC


  ###### FILE: Z6.10..DP1.2019-07-16T05.00.00.sac - 1
 ----------------------------------------

     CMPINC = 9.000000e+01

  ###### FILE: Z6.10..DP2.2019-07-16T05.00.00.sac - 2
 ----------------------------------------

     CMPINC = 9.000000e+01

  ###### FILE: Z6.10..DPZ.2019-07-16T05.00.00.sac - 3
 ----------------------------------------

     CMPINC = 1.800000e+02



### From 21-002 (SmartSolo SEG-D)
SAC> lh CMPINC


  ###### FILE: 9N.8064..GP1.2020-11-08T01.00.00.sac - 1
 ------------------------------------------

     CMPINC = 9.000000e+01

  ###### FILE: 9N.8064..GP2.2020-11-08T01.00.00.sac - 2
 ------------------------------------------

     CMPINC = 9.000000e+01

  ###### FILE: 9N.8064..GPZ.2020-11-08T01.00.00.sac - 3
 ------------------------------------------

     CMPINC = 1.800000e+02


### From 16-015 (RT130, broadband sensor)
SAC> lh CMPINC


  ###### FILE: YW.502.01.HH1.2016-06-25T01.00.00.sac - 1
 -------------------------------------------

     CMPINC = 9.000000e+01

  ###### FILE: YW.502.01.HH2.2016-06-25T01.00.00.sac - 2
 -------------------------------------------

     CMPINC = 9.000000e+01

  ###### FILE: YW.502.01.HHZ.2016-06-25T01.00.00.sac - 3
 -------------------------------------------

     CMPINC = 0.000000e+00


### From 17-034 (Texan)
SAC> lh CMPINC


  ###### FILE: 7B.12490..GPZ.2017-09-30T16.00.00.sac - 1
 -------------------------------------------

     CMPINC = 1.800000e+02


## Comparison with SAC files output from FDSNWS dataselect:
### From an experiment using a sensor with dip convention matching the SEG-D and Texan examples
SAC> lh CMPINC


  ###### FILE: Z9.M01..DH1.M.2019.010.120000.SAC - 1
 ---------------------------------------

     CMPINC = 9.000000e+01

  ###### FILE: Z9.M01..DH2.M.2019.010.120000.SAC - 2
 ---------------------------------------

     CMPINC = 9.000000e+01

  ###### FILE: Z9.M01..DHZ.M.2019.010.120000.SAC - 3
 ---------------------------------------

     CMPINC = 1.800000e+02

### From an experiment using a sensor with dip convention matching the example from 16-015:
SAC> lh CMPINC


  ###### FILE: 2B.PETZ..HHE.M.2019.296.120000.SAC - 1
 ----------------------------------------

     CMPINC = 9.000000e+01

  ###### FILE: 2B.PETZ..HHN.M.2019.296.120000.SAC - 2
 ----------------------------------------

     CMPINC = 9.000000e+01

  ###### FILE: 2B.PETZ..HHZ.M.2019.296.120000.SAC - 3
 ----------------------------------------

     CMPINC = 0.000000e+00

File: ph5/clients/ph5toms.py [hunk 1]
```diff
@@ -602,6 +602,13 @@ def create_trace(self, station_to_cut, mp=False):
                         self.ph5.get_receiver_t_by_n_i(stc.receiver_n_i)
                     azimuth = Receiver_t['orientation/azimuth/value_f']
                     dip = Receiver_t['orientation/dip/value_f']
+                    # dip is below horizontal axis
+                    # but SAC convention requires from vertical axis.
+                    dip = float(dip)
+                    if Receiver_t['orientation/description_s'] == 'Z':
```
Comment (timronan) [inline ph5/clients/ph5toms.py:608]: Why does the Z component control if the dip is 90 +dip vs 90-dip? Should all components be 90+dip or 90-dip? 

The SAC documentation says: 
"Component incident angle (degrees from upward vertical; SEED/MINISEED uses dip: degrees from horizontal down)."

Sorry for requesting clarification.

**Review by timronan — COMMENTED** [review]
Can you please address the inline comment. 

Thanks

File: ph5/clients/ph5toms.py (hunk 1 above)
Comment (damhuonglan) [reply ph5/clients/ph5toms.py:608]: I didn't look at the document of SAC, just follow the request from Holly which is stated above https://github.com/PIC-IRIS/PH5/pull/496#issuecomment-1018848379.
@hrotman-pic Holly, can you please help us to understand more about this? Thank you.

Comment (timronan) [reply ph5/clients/ph5toms.py:608]: I'm confused about the logic as well, but I am maybe missing something.

Comment (hrotman-pic) [reply ph5/clients/ph5toms.py:608]: If all components were 90-dip, CMPINC looks like this for a Fairfield SEG-D:

SAC> lh CMPINC


  ###### FILE: XE.221..DP1.2020-10-27T01.00.00.sac - 1
 -----------------------------------------

     CMPINC = 9.000000e+01

  ###### FILE: XE.221..DP2.2020-10-27T01.00.00.sac - 2
 -----------------------------------------

     CMPINC = 9.000000e+01

  ###### FILE: XE.221..DPZ.2020-10-27T01.00.00.sac - 3
 -----------------------------------------

     CMPINC = 0.000000e+00


CMPINC is dip+90 for Z channel on the version I approved so CMPINC will match the FDSNWS SAC examples (am assuming FDSNWS dataselect SAC is a reliable source), which is why I provided those examples.  Most PH5s experiments should look like the Z9.M01 FDSNWS example.  Rarely, one should look like the 2B.PETZ FDSNWS example.

@ascire-pic can elaborate, confirm, or deny if needed.

Comment (hrotman-pic) [reply ph5/clients/ph5toms.py:608]: Or, referring back to the SEED dip using a subset of the experiments I checked for testing & comparison.

20-001, XE.221.DPZ, SEED dip=90 (reversed convention): http://ds.iris.edu/mda/XE/221/--/DPZ/?starttime=2020-10-24&endtime=2020-11-25

SEED station Z9.M01.DHZ SEED dip=90 (reversed convention): http://ds.iris.edu/mda/Z9/M01/--/DHZ/?starttime=2019-01-04&endtime=2020-06-12

SEED dip is the same for these, so SAC CMPINC should be the same for these, and they are the same (CMPINC=180) on the current version.


16-015, YW.502.HHZ, SEED dip=-90: http://ds.iris.edu/mda/YW/502/01/HHZ/?starttime=2016-06-20&endtime=2016-10-30

SEED station 2B.PETZ.HHZ, SEED dip=-90: http://ds.iris.edu/mda/2B/PETZ/--/HHZ/?starttime=2019-10-21&endtime=2021-11-03

SEED dip is the same for these, so SAC CMPINC should also be the same for these, and they are (CMPINC=0) on the current version as well.

Does that also help?

Comment (timronan) [reply ph5/clients/ph5toms.py:608]: I think that the logic above is almost right but needs to have the signs flipped so :

PNCINC_Z = 90 - dip_Z
and 
PNCINC_H = 90 + dip_H

All of the components in the instrument are tied together at 90 degrees. If we have vertical components, PNCINC_Z = 90 - dip_Z, and horizontal components PNCINC_H = 90 + dip_H we'll keep this 90 degree relationship and we'll follow the `degrees from upward vertical` == clockwise convention. 

An example with the current logic described immediately above: 
 dip_Z = 85 
dip_H = 5 
meaning 
PNCINC_Z=5 
PNCINC_H = 95

What does everyone think?

Screen Shot 2022-01-28 at 10 56 06 AM

Comment (hrotman-pic) [reply ph5/clients/ph5toms.py:608]: This logic suggests that the CMPINC header is being written incorrectly on FDSN web service queries.

Comment (timronan) [reply ph5/clients/ph5toms.py:608]: Can you provide an example? Are we using a counter clockwise convention? I am, okay with whatever is decided.

Comment (hrotman-pic) [reply ph5/clients/ph5toms.py:608]: I have examples in my approving review, https://github.com/PIC-IRIS/PH5/pull/496#pullrequestreview-865103223, and link to the SEED dip for the one example of each in https://github.com/PIC-IRIS/PH5/pull/496#discussion_r794562682.  Please let me know if those don't make sense.  I realize my review is very long.

Comment (timronan) [reply ph5/clients/ph5toms.py:608]: -180 and 180 are the same position on the unit circle. The question is if we want our convention to be clockwise or counter clockwise.

Comment (timronan) [reply ph5/clients/ph5toms.py:608]: @hrotman-pic you are right, and I am off by a negative sign. I think we should merge the current logic.

**Review by timronan — APPROVED** [review]

**Pull request merged**